### PR TITLE
[Exp PyROOT] MultiPython - Build PyROOT with multiple Python versions

### PR DESCRIPTION
**Pull request by maxgalli**
Upgraded version of PR #4279.
Changes were necessary due to incompatibility with C++ modules.

Problem: due to how rootmap works, it's not possible to have multiple versions of a pair (shared library, pcm) with the same name, even in different directories. This feature lead to the impossibility of building multiple versions of libROOTPython, which used to contain, along with the pythonizations, also the TPython library.

Solution: TPython and pythonizations have been split, and are built in libROOTTPython and libROOTPythonizations respectively. We only have one version of the former (the last one built) and possibility to build multiple versions for the latter.

### CODE REVIEW

**Comment by maxgalli** [comment]
@phsft-bot build with flags -Dpyroot_experimental=ON

**Comment by maxgalli** [comment]
@phsft-bot build with flags -Dpyroot_experimental=ON

**Comment by maxgalli** [comment]
> Build failed on ROOT-ubuntu16/nortcxxmod.
> [See console output](https://epsft-jenkins.cern.ch/job/root-pullrequests-build/72624/console).
> 
> ### Errors:
> * FAILED: cd /mnt/build/workspace/root-pullrequests-build/build && /usr/bin/python /mnt/build/workspace/root-pullrequests-build/root/build/unix/makepchinput.py /mnt/build/workspace/root-pullrequests-build/root . ON etc/cling/Interpreter/DynamicExprInfo.h etc/cling/Interpreter/DynamicLookupRuntimeUniverse.h etc/cling/Interpreter/DynamicLookupLifetimeHandler.h etc/cling/Interpreter/Exception.h etc/cling/Interpreter/RuntimePrintValue.h etc/cling/Interpreter/RuntimeUniverse.h etc/cling/Interpreter/Value.h && /usr/local/bin/cmake -E env ROOTIGNOREPREFIX=1 /usr/bin/python /mnt/build/workspace/root-pullrequests-build/root/etc/dictpch/makepch.py etc/allDict.cxx.pch -I/mnt/build/workspace/root-pullrequests-build/build/include



Build and tests were tried separately in this machine.
The build directory is:

/home/sftnight/gallim4647/build

**Comment by maxgalli** [comment]
@phsft-bot build with flags -Dpyroot_experimental=ON

File: bindings/pyroot_experimental/pythonizations/CMakeLists.txt [hunk 1]
```diff
@@ -4,9 +4,9 @@
 # For the licensing terms see $ROOTSYS/LICENSE.
 # For the list of contributors see $ROOTSYS/README/CREDITS.
 
-############################################################################
-# CMakeLists.txt file for building ROOT bindings/pyroot_experimental package
-############################################################################
+###################################################################################
+# CMakeLists.txt file for building ROOT bindings/pyroot_experimental/PyROOT package
```
Comment (etejedor) [inline bindings/pyroot_experimental/pythonizations/CMakeLists.txt]: bindings/pyroot_experimental/pythonizations

File: bindings/pyroot_experimental/pythonizations/src/RDataFramePyz.cxx [hunk 2]
```diff
@@ -90,9 +90,9 @@ PyObject *PyROOT::MakeNumpyDataFrame(PyObject * /*self*/, PyObject * pydata)
    }
 
    // Create RDataFrame and build Python proxy
-   const auto err = gInterpreter->Declare("#include \"ROOT/RNumpyDS.hxx\"");
+   const auto err = gInterpreter->Declare("#include \"Pythonizations/RNumpyDS.hxx\"");
```
Comment (etejedor) [inline bindings/pyroot_experimental/pythonizations/src/RDataFramePyz.cxx]: The `RNumpyDS.hxx` header should still be inside `ROOT`

Comment (etejedor) [reply bindings/pyroot_experimental/pythonizations/src/RDataFramePyz.cxx]: This also needs fixing the directory where the header is placed, it should still be ROOT.

File: bindings/tpython/src/TPyClassGenerator.cxx [hunk 3]
```diff
@@ -0,0 +1,308 @@
+// Author: Enric Tejedor CERN  08/2019
```
Comment (etejedor) [inline bindings/tpython/src/TPyClassGenerator.cxx]: `TPyClassGenerator.cxx` was already there, it should not appear as a new file but rather as a renamed without changes?

File: bindings/pyroot_experimental/pythonizations/src/RDataFramePyz.cxx (hunk 2 above)
Comment (maxgalli) [reply bindings/pyroot_experimental/pythonizations/src/RDataFramePyz.cxx]: I just checked and it actually ends up in ROOT, I just put the wrong directory in this file

File: bindings/tpython/src/TPyClassGenerator.cxx (hunk 3 above)
Comment (maxgalli) [reply bindings/tpython/src/TPyClassGenerator.cxx]: I forgot to remove one of them, I'm fixing

File: bindings/pyroot_experimental/pythonizations/CMakeLists.txt [hunk 4]
```diff
@@ -25,7 +25,6 @@ set(py_sources
   ROOT/pythonization/_cppinstance.py
   ROOT/pythonization/_drawables.py
   ROOT/pythonization/_generic.py
-  ROOT/pythonization/_rbdt.py
```
Comment (etejedor) [inline bindings/pyroot_experimental/pythonizations/CMakeLists.txt]: Why removing this file?

File: bindings/pyroot_experimental/pythonizations/CMakeLists.txt [hunk 5]
```diff
@@ -42,7 +41,6 @@ set(py_sources
   ROOT/pythonization/_titer.py
   ROOT/pythonization/_tobject.py
   ROOT/pythonization/_tobjstring.py
-  ROOT/pythonization/_tree_inference.py
```
Comment (etejedor) [inline bindings/pyroot_experimental/pythonizations/CMakeLists.txt]: Why removing this?

File: bindings/pyroot_experimental/pythonizations/CMakeLists.txt [hunk 6]
```diff
@@ -70,58 +63,62 @@ ROOT_STANDARD_LIBRARY_PACKAGE(ROOTPython
...
 
 set(ROOTDirName python/ROOT)
-set(JupyROOTDirName python/JupyROOT)
-set(JsMVADirName python/JsMVA)
+set(Pyz_headers inc/Pythonizations)
+set(ROOT_headers inc/ROOT)
 
-ROOT_LINKER_LIBRARY(JupyROOT ${JupyROOTDirName}/src/IOHandler.cxx DEPENDENCIES Core CMAKENOEXPORT)
+# Add library libROOTPythonizations
+ROOT_LINKER_LIBRARY(ROOTPythonizations ${cpp_sources} DEPENDENCIES Core Tree cppyy CMAKENOEXPORT)
```
Comment (etejedor) [inline bindings/pyroot_experimental/pythonizations/CMakeLists.txt]: What is the purpose of `CMAKENOEXPORT` here?

File: bindings/pyroot_experimental/pythonizations/CMakeLists.txt [hunk 7]
```diff
@@ -70,58 +63,62 @@ ROOT_STANDARD_LIBRARY_PACKAGE(ROOTPython
...
-  DEPENDENCIES
-    Core
-    Tree
-    cppyy
 )
 
 set(ROOTDirName python/ROOT)
-set(JupyROOTDirName python/JupyROOT)
-set(JsMVADirName python/JsMVA)
+set(Pyz_headers inc/Pythonizations)
```
Comment (etejedor) [inline bindings/pyroot_experimental/pythonizations/CMakeLists.txt]: What is the reason for moving all the internal headers from `src` to `inc/Pythonizations` ?

File: bindings/pyroot_experimental/pythonizations/CMakeLists.txt [hunk 8]
```diff
@@ -70,58 +63,62 @@ ROOT_STANDARD_LIBRARY_PACKAGE(ROOTPython
...
 
-#---Install python modules----------------------------------------------------------------------
+# Copy pure Python modules
+file(COPY ${ROOTDirName} DESTINATION ${py_localruntimedir})
+install(DIRECTORY ${ROOTDirName} DESTINATION ${CMAKE_INSTALL_LIBDIR})
 
-file(COPY ${ROOTDirName} DESTINATION ${localruntimedir})
-install(DIRECTORY ${ROOTDirName} DESTINATION ${runtimedir})
+# Copy headers inside build_dir/include/Pythonizations
+file(COPY ${Pyz_headers}/ DESTINATION ${CMAKE_BINARY_DIR}/include/Pythonizations)
```
Comment (etejedor) [inline bindings/pyroot_experimental/pythonizations/CMakeLists.txt]: We don't need these headers in the build directory, do we?

File: bindings/pyroot_experimental/pythonizations/setup.py.in [hunk 9]
```diff
@@ -0,0 +1,9 @@
+from distutils.core import setup
+
+setup(name = 'ROOT',
+      version = '${ROOT_VERSION}',
+      packages=['ROOT', 'ROOT.pythonization'],
+      package_dir={'' : '${CMAKE_CURRENT_SOURCE_DIR}/python'},
+      url = 'https://root.cern.ch',
+      author_email = 'root-dev@cern.ch',
```
Comment (etejedor) [inline bindings/pyroot_experimental/pythonizations/setup.py.in]: You can put your e-mail here! :)

File: bindings/tpython/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  SOURCES
+    src/TPyArg.cxx
+    src/TPyClassGenerator.cxx
+    src/TPyReturn.cxx
+    src/TPython.cxx
+  DICTIONARY_OPTIONS
+    -writeEmptyRootPCM
+  DEPENDENCIES
+    Core
+    Tree
```
Comment (etejedor) [inline bindings/tpython/CMakeLists.txt]: Does `TPython` depend on `Tree`?

File: bindings/tpython/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    cppyy
+)
+
+# Disables warnings caused by Py_RETURN_TRUE/Py_RETURN_FALSE
+if(NOT MSVC)
+    target_compile_options(ROOTTPython PRIVATE -Wno-strict-aliasing)
```
Comment (etejedor) [inline bindings/tpython/CMakeLists.txt]: This is not necessary for `TPython`. `Py_RETURN_TRUE/FALSE` are not used.

File: bindings/tpython/src/TPython.cxx [hunk 12]
```diff
@@ -167,7 +162,7 @@ Bool_t TPython::Import(const char *mod_name)
 
    // allow finding to prevent creation of a python proxy for the C++ proxy
    Py_INCREF(mod);
-   PyModule_AddObject(PyROOT::gRootModule, mod_name, mod);
+   PyModule_AddObject(CPyCppyy::PyStrings::gThisModule, mod_name, mod);
```
Comment (etejedor) [inline bindings/tpython/src/TPython.cxx]: `PyROOT::gRootModule` is a Python module, but `CPyCppyy::PyStrings::gThisModule` is a Python string. This replacement is therefore wrong, we should put `CPyCppyy::gThisModule`.

File: bindings/tpython/src/TPython.cxx [hunk 13]
```diff
@@ -92,10 +91,6 @@
 ClassImp(TPython);
 static PyObject *gMainDict = 0;
 
-namespace PyROOT {
-R__EXTERN PyObject *gRootModule;
```
Comment (etejedor) [inline bindings/tpython/src/TPython.cxx]: You will need to do this with `gThisModule` of cppyy:
```
namespace CPyCppyy {
    extern PyObject* gThisModule;
}
```

File: bindings/tpython/src/TPython.cxx [hunk 14]
```diff
@@ -12,7 +12,6 @@
 // Bindings
 #include "CPyCppyy.h"
 #include "PyStrings.h"
-#include "PyROOTStrings.h"
 #include "TPython.h"
```
Comment (etejedor) [inline bindings/tpython/src/TPython.cxx]: I would place this include the first, before the `//Bindings` section

File: bindings/tpython/src/TPython.cxx [hunk 15]
```diff
@@ -404,7 +399,7 @@ const TPyReturn TPython::Eval(const char *expr)
       return TPyReturn(result);
 
    // explicit conversion for python type required
-   PyObject *pyclass = PyObject_GetAttr(result, PyROOT::PyStrings::gClass);
+   PyObject *pyclass = PyObject_GetAttr(result, CPyCppyy::PyStrings::gBase);
```
Comment (etejedor) [inline bindings/tpython/src/TPython.cxx]: `gBase` is `"__base__"` and `gClass` is `"__class__"`. Why this change?

Comment (etejedor) [reply bindings/tpython/src/TPython.cxx]: If the problem is that `"__class__"` is not in the PyStrings of cppyy, you can just do:
```
PyObject *pyclass = PyObject_GetAttrString(result, const_cast<char*>("__class__"));
```

**Comment by etejedor** [comment]
@phsft-bot build with flags -Dpyroot_experimental=OFF
